src/newcharacter.cpp

```diff
@@ -1088,8 +1070,61 @@ int set_profession(WINDOW *w, player *u, int &points)
     ctxt.register_action("NEXT_TAB");
     ctxt.register_action("SORT");
     ctxt.register_action("HELP_KEYBINDINGS");
+    ctxt.register_action("FILTER");
+
+    bool recalc_profs = true;
+    int profs_length;
+    std::string filterstring = "";
+    std::vector<const profession *> sorted_profs;
 
     do {
+        if (recalc_profs) {
+            sorted_profs.clear();
+            for (profmap::const_iterator iter = profession::begin(); iter != profession::end(); ++iter) {
```

Comment:
How about a static member like profession::all that happens to be just an
instance of profession? This could also be extended to support things like
`for( auto &prof : profession::filtered(some_filter) )`.

On Wed, Aug 26, 2015 at 2:07 PM, BevapDin <notifications@github.com> wrote:

> In src/newcharacter.cpp
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13413#discussion_r37969796>
> :
>
> >
> >      do {
> > +        if (recalc_profs) {
> > +            sorted_profs.clear();
> > +            for (profmap::const_iterator iter = profession::begin(); iter != profession::end(); ++iter) {
>
> Well, that is disappointing. profession and scenario expose the begin and
> end iterator via *static* functions, and the syntax for the range-based
> for loop is for( range_declaration : range_expression ) and
> range_expression must be an object. for( auto p : profession ) doesn't
> work because range_expression would be a type, not an object.
>
> You could do this (please don't, it works, but it's very un-intuitive):
>
> for( auto &prof : profession() ) {
>
> This works because the temporary object profession() has begin and end
> methods which return iterators (through they are static, but that doesn't
> matter, profession().begin() is still valid).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13413/files#r37969796>.
>